### ISSUE TITLE

        
        Result Message:	OneTimeSetUp: Category name must not contain ',', '!', '+' or '-'
      

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit#1678]
@Andrey-Bushman commented on [Fri Jul 15 2016](https://github.com/nunit/nunit3-vs-adapter/issues/206)

I want to group my tests on the base of tested method signature (one test checks return values, and also exist the individual tests for each `out` parameter checking). For getting this result I use the `Category` attribute:

`[Category("LocalizedProduct.TryParse(String, out ILocalizedProduct)")]`

But tests which was marked by such attribute value are failed with the message:

> Result Message:   OneTimeSetUp: Category name must not contain ',', '!', '+' or '-'

Why this restriction exists?

I could be use this variant:

`[Category("LocalizedProduct.TryParse(String | out ILocalizedProduct)")]`

But the comma using would be _more legible_ for me at this case.

---

@CharliePoole commented on [Fri Jul 15 2016](https://github.com/nunit/nunit3-vs-adapter/issues/206#issuecomment-232997494)

We don't allow characters that have a special meaning in NUnit TSL (Test Selection Language). See https://github.com/nunit/docs/wiki/Test-Selection-Language

It may be that the exclusion of the comma is obsolete - I'll have to check further.

---

@CharliePoole commented on [Fri Jul 15 2016](https://github.com/nunit/nunit3-vs-adapter/issues/206#issuecomment-233078139)

An earlier version supported

```
--where "cat == A,B,C"
```

in place of the longer 

```
--where "cat==A||cat==B||cat==C"
```

We have an outstanding issue asking us to restore the shorter syntax. We would have to decide about that issue before it could be possible to change this.

I'm going to go ahead and move this issue to the nunit project where it belongs.

**Comment by ChrisMaddock** [comment]
@CharliePoole - this "outstanding issue" you mention, I can't find it, so assume it was resolved? Can we unblock this one and mark if up for someone to resolve?

**Comment by ChrisMaddock** [comment]
Also, am I correct that both the comma and hyphen no longer feature in the Test Selection Language? Do we also need to add some other characters which now do?

**Comment by CharliePoole** [comment]
@ChrisMaddock The issue was #925, which was closed as won't fix. We should have unblocked this.

To your second question, I don't __think__ we do but it should be checked. IIRC, we allow the arg to be in quotes, which TSL defines as `"..."`, `'...'` or `/.../`. So it should be possible to use pretty much anything in the name of a category. However, the implementor should try all possible special chars in tests to be sure.
